src/compiler/scala/tools/nsc/symtab/classfile/JavaSignatureWalker.scala

```diff
@@ -0,0 +1,108 @@
+package scala.tools.nsc.symtab.classfile
+
+import scala.reflect.internal.ClassfileConstants._
+
+abstract class JavaSignatureWalker {
+  def visitName(internalName: CharSequence): Unit
+  def raiseError(msg: String): Unit
+
+  def walk(sig: CharSequence): Unit = {
+    def raiseError(msg: String): Unit = {
+      this.raiseError(s"$msg (while parsing $sig)")
+    }
+    var index = 0
+    val end = sig.length
+    def accept(ch: Char) {
+      assert(current() == ch, (current(), ch))
+      index += 1
+    }
+    def skipSubName(isDelimiter: Char => Boolean): Unit = {
+      val start = index
+      while (!isDelimiter(current())) { index += 1 }
```

Comment:
If `current()` doesn't stop / throw after `end`, it might be better to check for `\0` to prevent an infinite loop.